plutarch-extra/Plutarch/Extra/List.hs

```diff
@@ -8,16 +28,264 @@ preverse =
+
+{- | / O(n) /. Merge two lists which are assumed to be ordered, given a custom comparator.
+    The comparator should return true if first value is less than the second one.
+-}
+pmergeBy ::
+  forall (a :: S -> Type) (s :: S) list.
+  (PIsListLike list a) =>
+  Term
+    s
+    ( (a :--> a :--> PBool)
+        :--> list a
+        :--> list a
+        :--> list a
+    )
+pmergeBy =
+  phoistAcyclic $
+    pfix #$ plam $ \self comp a b ->
+      pelimList
+        ( \x xs ->
+            pelimList
+              ( \y ys ->
+                  pif
+                    (comp # x # y)
+                    (pcons # x #$ self # comp # xs # b)
+                    (pcons # y #$ self # comp # a # ys)
+              )
+              a
+              b
+        )
+        b
+        a
+
+{- | / O(n log n) /. Merge sort, bottom-up version, given a custom comparator.
+
+   Assuming the comparator returns true if first value is less than the second one,
+    the list elements will be arranged in ascending order, keeping duplicates in the order
+    they appeared in the input.
+-}
+pmsortBy ::
+  forall s a l.
+  (PIsListLike l a, PIsListLike l (l a)) =>
+  Term s ((a :--> a :--> PBool) :--> l a :--> l a)
+pmsortBy = phoistAcyclic $
+  plam $ \comp xs ->
+    mergeAll # comp # (pmap # psingleton # xs)
+  where
+    mergeAll :: Term _ ((a :--> a :--> PBool) :--> l (l a) :--> l a)
+    mergeAll = phoistAcyclic $
+      pfix #$ plam $ \self comp l ->
+        pelimList
+          ( \x xs ->
+              pif
+                (pnull # xs)
+                x
+                (self # comp #$ mergePairs # comp # l)
+          )
+          pnil
+          l
+    mergePairs :: Term _ ((a :--> a :--> PBool) :--> l (l a) :--> l (l a))
+    mergePairs = phoistAcyclic $
+      pfix #$ plam $ \self comp l ->
+        pelimList
+          ( \x xs ->
+              pelimList
+                ( \y ys ->
+                    pcons # (pmergeBy # comp # x # y) # (self # comp # ys)
+                )
+                l
+                xs
+          )
+          pnil
+          l
+
+-- | A special case of 'pmsortBy' which requires elements have 'POrd' instance.
+pmsort ::
+  (POrd a, PIsListLike l a, PIsListLike l (l a)) =>
+  Term s (l a :--> l a)
+pmsort = phoistAcyclic $ pmsortBy # comp
+  where
+    comp = phoistAcyclic $ plam (#<)
+
+-- | /O(n log n)/. Remove dupicate elements in a list with the given function to decide duplicate.
+pnubBy ::
+  forall (a :: S -> Type) (s :: S) list.
+  (PIsListLike list a, PIsListLike list (list a)) =>
+  Term
+    s
+    ( (a :--> a :--> PBool)
+        :--> list a
+        :--> list a
+    )
+pnubBy = phoistAcyclic $
+  plam $ \eq l ->
+    pelimList (\x xs -> go # eq # x # xs) l l
+  where
+    go = phoistAcyclic pfix #$ plam $ \self eq seen l ->
+      pelimList
+        ( \x xs ->
+            pif
+              (eq # x # seen)
+              (self # eq # seen # xs)
+              (pcons # seen #$ self # eq # x # xs)
```

Comment:
You're fixing at the wrong place. You want to fix after applying the equality function.